Replace LB-IPAM IP allocator

dylandreimerink opened the pull request:
This PR introduces a new IP allocator, replaces the existing IP allocator used by LB-IPAM with this new allocator. This removes the limitation on IPv6 ranges, the full range can now be used (only limited by the resources required to book keep). It also removes the reservation of the base and broadcast IP address making it possible to allocate single IPs. Lastly, support is now added for non-CIDR IP ranges such as x.x.x.100-x.x.x.200.

Fixes: #24351
Fixes: #22005
Fixes: #28255

```release-note
Replace LB-IPAM IP allocator to remove limitations and enable additional features
```

dylandreimerink committed:
lbipam: Don't reserve IPs for CIDRs with 1 or 2 IPs

LB-IPAM was still reserving all IPs in a /32 or /31 IPv4 CIDR if
`allowFirstLastIPs` is `no`. This behavior doesn't make any sense, so
this commit makes it so we only apply this behavior for CIDRs with more
than two IPs.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
lbipam,docs: Added docs for LB-IPAM reserved IPs

Added a new chapter to the LB-IPAM docs to explain the reservation
of the first and list IPs of CIDRs and how to allow them to be used.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

ysksuzuki commented on Documentation/network/lb-ipam.rst:
```diff
@@ -50,11 +55,26 @@ After adding the pool to the cluster, it appears like so.
+these by default. Exceptions are /32 and /31 IPv4 CIDRs and /128 and /127 IPv6
+CIDRs since these only have 1 or 2 IPs respectively.
+
+If you wish to use these dodgy IPs, you can set the ``.spec.allowDodgyIPs`` field
```
If we configure an expensive address pool such as global ipv4 addresses, we must set `allowDodgyIPs=true`, correct? If so, I'm not sure this spec makes sense because there are more cases where you want to set `allowDodgyIPs=false` from my experience. Plus, I'm confused because it's the opposite of the MetalLB `AvoidBuggyIPs` flag. 

We use `/28` for global and bastion networks. So I assume we need to set `allowDodgyIPs=true` accordingly.

dylandreimerink replied:
> If we configure an expensive address pool such as global ipv4 addresses, we must set allowDodgyIPs=true, correct? If so, I'm not sure this spec makes sense because there are more cases where you want to set allowDodgyIPs=false from my experience.

Yea, that is correct, if you want to use the first and last IP of the global range then you have to set `allowDodgyIPs` to `true`. If unspecified, the value defaults to `false`, so unless you specify `allowDodgyIPs=true`, the first and last IPs are reserved, the same behavior we currently have.

> I'm confused because it's the opposite of the MetalLB `AvoidBuggyIPs` flag.

Yea, this can't really be helped because we need to maintain backwards compatibility for existing policies, otherwise we may break setups on upgrade.

> We use /28 for global and bastion networks. So I assume we need to set allowDodgyIPs=true accordingly.

Practical example for `10.0.0.128/28`, if set to `true`, the allocate range is `10.0.0.128-10.0.0.143`, if `false` or unspecified its `10.0.0.129-10.0.0.142`. So you should pick which is desirable for your scenario

ysksuzuki replied:
Thank you for your comment!

I'm worried that users migrating from MetalLB(either MetalLB BGP ControlPlane or standalone) will be confused. With that said, it is understandable that backward compatibility must be maintained.

Wouldn't it be acceptable to break compatibility after reminding users in the upgrade guide in this case?

dylandreimerink replied:
> Wouldn't it be acceptable to break compatibility after reminding users in the upgrade guide in this case?

It could be in some cases, but consider the worst scenario for this case if a user doesn't pay attention or misses the notice. Any user with existing pools will suddenly be able to allocate the dodgy IPs, hard to notice if you are not looking for it. Lets say a service gets assigned a broadcast IP in a network which has this feature enabled. At best the feature doesn't work, at worst it causes a DoS or security issue.

I think the confusion concern is valid, but I hope it isn't as big a deal as we might think. I would like to know what others think.

ysksuzuki replied:
I think I'm slightly biased because I'm the user planning to migrate from MetalLB. So yeah, we should ask what others think.

networkop replied:
I would change this behaviour and enable this by default. Here's my rationale:
network and broadcast addresses only make sense on a network segment. When we allocate and announce individual host prefixes, they have no notion of broadcast or network addresses. 
For example, 192.168.255.255 may be a broadcast address for a segment with prefix 192.168.0.0/16 but may be a normal address for prefix 192.168.0.0/15. 
Even if we announce individual host IPs and they get aggregated on a border router, they still are not assigned to any network segment (aggregate route is configured as null-route).

The only caveat may be with L2 announcer where, if the range is picked incorrectly, allocated IP can overlap with local subnet's broadcast address. While this is more of an operational problem, I don't think this will result in any loss of connectivity too.

---

I've got an extra ask, can we please not use the word `dodgy`? This is not a generally-used term and may get a lot of people confused. I can suggest `allowFirstLastlIPs` or `prefixAsRange` or `CIDRasRange`.

dylandreimerink replied:
I brought this issue up in last weeks Community meeting to ask for advice regarding breakage of backwards compatibility. I think we settled on a decent middle ground, that is, in 1.15 we will not break anything, introduce a setting to change the behavior. Then next release, 1.16 we will switch over the default.

We will add warnings in the feature docs and upgrade guide, warning people of the upcoming change, and asking people to explicitly set what they want. That way, we have an upgrade path where we don't break setups on upgrade and change the behavior even if it takes a bit.

I am thinking of changing the setting from a boolean to a enum, so we get something like `allowFirstLastIPs: yes|no|default` with `default` being the value chosen if the field is not specified. This also should make the field different enough from MetalLB's implementation to not cause confusion.

Would this be acceptable?

ysksuzuki replied:
@dylandreimerink Thank you for bringing this up!

So, in 1.15, we will introduce `allowFirstLastIPs=false` by default to ensure backward compatibility, and in 1.16 we will change its default, which is  `allowFirstLastIPs=true` by default. Is my understanding correct?

dylandreimerink replied:
Yes. Currently there is no setting, so if we add the setting and change the default, then users have no chance to set the setting before the default switches. So by doing it in two versions, we allow users some time to explicitly set what they want the behavior to be for existing pools before we switch.

ysksuzuki replied:
I understand. It makes sense to me. Thanks!

dylandreimerink replied:
Alright, should be good now

ysksuzuki reviewed (approved):
LGTM. Thanks!

joestringer commented on pkg/ipalloc/ipalloc.go:
Did we have a discussion about whether this IP allocator might overlap with some other IP allocators we have in the tree?

This file should also be added to CODEOWNERS under @cilium/sig-ipam .

dylandreimerink replied:
I don't think we have, but I am happy to

christarazi replied:
Coming from sig-ipam, have we looked into how we could make this implementation backwards compatible with the existing users of `pkg/ipam/service/ipallocator` so that all users are using one IPAM implementation? It sounds like this new implementation can be a superset, is that right?

dylandreimerink replied:
Hmm, so here are the two interfaces:

ipallocator:
```
type Interface interface {
	Allocate(net.IP) error
	AllocateNext() (net.IP, error)
	Release(net.IP) error
	ForEach(func(net.IP))
	CIDR() net.IPNet
	Has(ip net.IP) bool
}
```

New LB-IPAM allocator:
```
type Allocator[T any] interface {
	AllocAny(val T) (netip.Addr, error)
	Alloc(ip netip.Addr, val T) error
	Update(ip netip.Addr, val T) error
	Free(ip netip.Addr) error
	Get(ip netip.Addr) (T, bool)
	Stats() (allocated uint64, available *big.Int)
	Range() (from, to netip.Addr)
}
```

The additional features of the LB-IPAM allocator require certain changes in the API. But I think we can make a small adapter to go from the old interface to this one. Its mostly converting between types and placing some restrictions during construction.

```
Interface.Allocate(net.IP) error         -> Allocator.Alloc(ip netip.Addr, val T) error
Interface.AllocateNext() (net.IP, error) -> Allocator.AllocAny(val T) (netip.Addr, error)
Interface.Release(net.IP) error          -> Allocator.Free(ip netip.Addr) error
Interface.ForEach(func(net.IP))          -> // this one would require some modification, but can be done
Interface.CIDR() net.IPNet               -> Allocator.Range() (from, to netip.Addr)
Interface.Has(ip net.IP) bool            -> Allocator.Get(ip netip.Addr) (T, bool)
```

dylandreimerink replied:
I added an adapter which converts from the old to the new interface, would be great if you can take a look to see if this is adequate.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+)
+
+// Allocator allocates IP addresses
+type Allocator[T any] interface {
```
Could you be a bit more specific about the intention of the type variable here, I found that a little confused when first reading it.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+
+	// instead of a for {} loop, pick a unreasonably high iteration count and return an error
+	// when reaching it to prevent infinite loops in case of bugs
+	const maxIter = 1_000_000
```
could we take the minimum of this and the max size of the allocator to optimize a bit?

dylandreimerink replied:
So you want to base this max iteration count on the side of the list? Something like `maxIter = len(*abl) + someMargin`?

tommyp1ckles replied:
Yeah was thinking that instead of setting some arbitrarily large max_iter we could find some upper bound based on the allocator size.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+// This coalesces what otherwise would be multiple sort operations if lots of allocations happen in a short
+// time. Sorting of available blocks improves memory usage in the long term but is not critical to the
+// operation of the allocator, and thus its alright to delay the sort.
+func (a *HashAllocator[T]) scheduleSort() {
```
Just a thought, perhaps we could use a priority queue to maintain these as an ordered set with lg(n) insertions

dylandreimerink replied:
Yea, a priority queue indeed sounds like a better idea. That should allow us to eliminate this sorting routine which I didn't like to begin with. I will dedicate some time to try that out.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+	// At this point, it must be an IP in the middle of the block, so we need to split it
+
+	// Remove the current block from the list
+	*abl = slices.Delete(*abl, index, index+1)
```
This seems potentially expensive for large, fragmented sets.

dylandreimerink replied:
Yea, it could be. Though the list only contains pointers, so its just a copy of 8*#blockRightOfIndex amount of bytes, you would have to have things fragmented very heavily. But if we go for a heap to sort blocks this operation should become cheaper only a pointer update.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+		block: block{
+			from: availableBlock.from,
+			to:   ip.Prev(),
+			size: addrDiff(availableBlock.from, ip.Prev()),
```
nit: we could use calcSize here to keep things consistent.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+
+var _ Allocator[bool] = (*HashAllocator[bool])(nil)
+
+// NewHashAllocator creates a new IP allocator which will allocate IPs between `start` and `stop`.
```
Nit: Might be helpful to readers to mention that this is an inclusive set on both ends of the range.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+	if lb.prev != nil {
+		fmt.Fprintf(&sb, "(%s - %s) <- ", lb.prev.from, lb.prev.to)
+	}
+	fmt.Fprintf(&sb, "(%s - %s)[%s]", lb.from, lb.to, lb.bigIntSize().String())
```
Not sure if this is right, when I do something like:

```go
	alloc, err := NewHashAllocator[bool](
		netip.MustParseAddr("10.0.0.0"),
		netip.MustParseAddr("10.0.0.100"),
		10,
	)
	err = alloc.Alloc(netip.MustParseAddr("10.0.0.1"), true)
	fmt.Println(alloc.availableBlocks)
```
I get the output:

```
linked-block-list[
(10.0.0.0 - 10.0.0.0)[0] -> (10.0.0.2 - 10.0.0.100)
(10.0.0.0 - 10.0.0.0) <- (10.0.0.2 - 10.0.0.100)[98]
]
```

I think the size on the left side should be 1, not 0

dylandreimerink replied:
Yes, its somewhat subtle but if 0 and 100 are allocatable, then 0-100 contains 101 usable IPs. The 98 in the .2-.100 should also be 99 IPs. I have corrected the off by one now.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+// The `start` IP should be lower then or equal to `stop`. The `caphint` is passed to the backing
+// hashmap to save on map resizing when initial IP count is broadly known.
+func NewHashAllocator[T any](start, stop netip.Addr, caphint int) (*HashAllocator[T], error) {
+	if start.BitLen() == 0 {
```
This returns no error if you pass `netip.MustParseAddr("0.0.0.0")`, perhaps we'd be better off doing ` isValid()`?

tommyp1ckles commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -640,26 +641,26 @@ func (ipam *LBIPAM) stripOrImportIngresses(sv *ServiceView) (statusModified bool
-		if ip != nil {
+		ip, err := netip.ParseAddr(svc.Spec.LoadBalancerIP)
+		if err == nil {
 			ips = append(ips, ip)
```
we should probably log if these are failing to parse.

tommyp1ckles commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -1089,8 +1089,13 @@ func (ipam *LBIPAM) handleNewPool(ctx context.Context, pool *cilium_api_v2alpha1
+
+			// If the first and last IPs are the same or adjacent, we would reserve the entire range.
+			// Only reserve first and last IPs for ranges /30 or /126 and larger.
+			if !(from.Compare(to) == 0 || from.Next().Compare(to) == 0) {
```
I'm wondering, if this is meant to be exclusive at both ends by default, should we just consider using `allowFirstLastIPs`when the ranges are /32?

dylandreimerink replied:
I don't understand the question sorry. This piece of code is to always allow first and last IP for /32 and /31 ranges because reserving the full range is always nonsensical

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
 	//
 	// +kubebuilder:validation:Optional
 	ServiceSelector *slimv1.LabelSelector `json:"serviceSelector"`
+	// AllowFirstLastIPs, if set to yes means that the first and list IPs of each CIDR will be allocatable.
```
s/list/last/

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
 	// +kubebuilder:validation:Optional
 	ServiceSelector *slimv1.LabelSelector `json:"serviceSelector"`
+	// AllowFirstLastIPs, if set to yes means that the first and list IPs of each CIDR will be allocatable.
+	// If no, these IPs will be reserved.
```
s/no, these IPs/unspecified, the IPs defined in cidrs or blocks/

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
-	Cidrs []CiliumLoadBalancerIPPoolCIDRBlock `json:"cidrs"`
+	// +kubebuilder:validation:Optional
+	Cidrs []CiliumLoadBalancerIPPoolIPBlock `json:"cidrs"`
+	// CiliumLoadBalancerIPPoolCIDRBlock is a list of CIDRs comprising this IP Pool
```
Why are two fields, `cidrs` and `blocks`, the same type with the same documented behavior? Instead, `cidrs` can continue to represent CIDR notation and then add `spec.ranges` that defines a list of IP ranges. For example:

```go
// CiliumLoadBalancerIPPoolIPRange describes a range of IP addresses.
type CiliumLoadBalancerIPPoolIPRange struct {
        // Start defines the first IP address of an IP pool range.
	// +required
	Start string `json:"start"`
        // Stop defines the last IP address of an IP pool range.
	// +required
	Stop string `json:"stop"`
}
```
Thoughts?

dylandreimerink replied:
> Why are two fields, cidrs and blocks, the same type with the same documented behavior?

We had `cidrs` before, but the idea is to rename it do `blocks`. So we currently have 2 fields which are aliased for backwards compatibility. 

> Instead, cidrs can continue to represent CIDR notation and then add spec.ranges that defines a list of IP ranges

Yes, that would have also worked. I think it comes to do style/preference.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=yes;no
+	AllowFirstLastIPs string `json:"allowFirstLastIPs,omitempty"`
```
Why not use a `bool` type... are other enums expected in the future?

dylandreimerink replied:
Because the value is optional, this provides a tri-state type which allows us to preserve existing behavior for users who are upgrading with an existing policy. This could also be done with a `*bool`, however when I brought this up in the community meeting this was preferred, also seems to be more inline with https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types

danehans replied:
@dylandreimerink thanks for the clarification. Note that the API conventions also recommend using string-type aliases instead of enums. For example:

```go
{
...
    AllowFirstLastIPs AllowFirstLastIPType `json:"allowFirstLastIPs,omitempty"`

}

type AllowFirstLastIPType string

const (
    NoAllowFirstLastIPType AllowFirstLastIPType = "No"
    YesAllowFirstLastIPType AllowFirstLastIPType = "Yes"
)
```

dylandreimerink replied:
That is a good point, thanks, I added string-type aliases :+1:

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=yes;no
+	AllowFirstLastIPs string `json:"allowFirstLastIPs,omitempty"`
 	// CiliumLoadBalancerIPPoolCIDRBlock is a list of CIDRs comprising this IP Pool
```
s/CiliumLoadBalancerIPPoolCIDRBlock/Cidrs/

danehans commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -991,14 +994,14 @@ func (ipam *LBIPAM) allocateIPAddress(
 	}
 
-	return nil, nil, nil
+	return netip.Addr{}, nil, nil
```
Why not define `newIP` as a pointer if the func returns an empty value?

dylandreimerink replied:
The empty/uninitialized IP is an intended feature of the package in combination with the `IsValid` method. Always passing around values and not pointer avoids a number of foot guns such as nil dereferences and modification of shared values.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,20 @@ type CiliumLoadBalancerIPPoolSpec struct {
+	// CiliumLoadBalancerIPPoolCIDRBlock is a list of CIDRs comprising this IP Pool
+	//
+	// +kubebuilder:validation:Optional
+	Blocks []CiliumLoadBalancerIPPoolIPBlock `json:"blocks"`
```
Since the doc states "A block can be specified with CIDR
notation (<prefix>/<bits>) or a range notation with a start and stop IP.", consider making CiliumLoadBalancerIPPoolSpec a [union type](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions) and introducing a union discriminator to discriminate among supported allocation types. Using a union type with a discriminator can make validation and code clarity easier.

Example:
```go
// CiliumLoadBalancerIPPoolSpec is a human readable description for
// a load balancer IP pool.
// +union
type CiliumLoadBalancerIPPoolSpec struct {
  // AllocType defines the type of IP allocation to use for this pool.
  // Supported types are:
  //
  // "CIDR": An IP allocator based on CIDR notation, e.g. 1.1.1.0/24.
  // "Range": An IP allocator based on IP range notation, e.g. 1.1.1.1-1.1.1.10.
  //
  // If unset, defaults to CIDR.
  //
  // +unionDiscriminator
  // +required
  // +kubebuilder:default=CIDR
  // +kubebuilder:validation:Enum=CIDR,Range
  AllocType PoolAllocType `json:"allocType"`
  // CiliumLoadBalancerIPPoolCIDRBlock is a list of CIDRs that comprise this IP Pool.
  //
  // This field is only applicable if type is "CIDR".
  //
  // +kubebuilder:validation:Optional
  Cidrs []CiliumLoadBalancerIPPoolIPBlock `json:"cidrs"`
  // CiliumLoadBalancerIPPoolIPRange is a list of IP ranges that comprise this IP Pool.
  //
  // This field is only applicable if type is "Range".
  //
  // +kubebuilder:validation:Optional
  Blocks []CiliumLoadBalancerIPPoolIPRange `json:"ranges"`
  ...
}

// PoolAllocType is a type of IP pool allocation.
// +kubebuilder:validation:Enum=CIDR,Range
type PoolAllocType string

const (
  // CIDRPoolAllocType is an IP allocation type based on CIDR notation.
  CIDRPoolAllocType PoolAllocType = "CIDR"
  // RangePoolAllocType is an IP allocation type based on IP range notation.
  RangePoolAllocType PoolAllocType = "Range"
)
```

For backwards compatible upgrades, the union discriminator can instead be `+optional` or implement logic that adds this field to existing objects that were created with the previous version of this CRD.

dylandreimerink replied:
Perhaps we need to rephrase the docs. But I would like to keep the ability to mix CIDR and range notation in the same spec like:
```
spec:
  blocks:
  - cidr: "10.0.12.0/24"
  - start: "10.0.13.5"
    stop: "10.0.13.8"
```

I see no point in artificially limiting to one or the other.

danehans commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -820,7 +823,7 @@ func (ipam *LBIPAM) satisfyService(sv *ServiceView) (statusModified bool, err er
 		// If the allocated IP isn't found in the assigned list, assign it
 		if slices.IndexFunc(sv.Status.LoadBalancer.Ingress, func(in slim_core_v1.LoadBalancerIngress) bool {
-			return net.ParseIP(in.IP).Equal(alloc.IP)
+			return netip.MustParseAddr(in.IP).Compare(alloc.IP) == 0
```
According to the [docs](https://pkg.go.dev/net/netip#MustParseAddr), MustParseAddr should only be used in tests.

danehans commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,617 @@
+		hi: addrStruct.FieldByName("hi").Uint(),
+		lo: addrStruct.FieldByName("lo").Uint(),
+	}
+}
```
Would it be more performant and maintainable to use the `encoding/binary` pkg? For example:
```go
func uint128FromAddr(addr netip.Addr) uint128 {
	bytes := addr.As16()
	return uint128{
		hi: binary.BigEndian.Uint64(bytes[:8]),
		lo: binary.BigEndian.Uint64(bytes[8:]),
	}
}
```

dylandreimerink replied:
Yes! Did a quick benchmark, this goes from `167.0 ns/op` to `7.397 ns/op`, did not expect that, thanks!

dylandreimerink committed:
ipalloc: Added new IP allocator for LB-IPAM update

Currently LB-IPAM uses cilium/ipam/service/ipallocator to allocate its
IPs. While this is a decent allocator for most purposes, it is holding
LB-IPAM back in certain aspects.

This allocator aims to address the limitations encountered with the
current allocator such as:

* The new allocator supports non-CIDR ranges (x.x.x.100-x.x.x.200)
* The new allocator does not reserve the base and broadcast IPs
* The new allocator is not limited to 65536 IPs for IPv6
* The new allocator can associate arbitrary data with IPs via generics
  not just a boolean in-use or not-in-use

The removal of these limitations should allow us to make LB-IPAM more
user friendly and to implement some additional features.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
lbipam: Replace cilium/ipam with pkg/ipalloc in LB-IPAM

This commit replaces the underlying IP allocator used by LB-IPAM from
the cilium/ipam/service/ipallocator package to the new pkg/ipalloc
package.

This is just a 1-to-1 replacement without making use of any of the new
features of this allocator. The only change in behavior is that the
base and broadcast addresses are no longer reserved by default.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
lbipam: Allow range notation on CiliumLoadBalancerIPPool

Now that we have the new IP allocator, we can allocate non-CIDR ranges.
The ideal new syntax looks like:

```
[...]
spec:
  blocks:
  - cidr: "10.0.0.0/24"
  - start: "20.0.0.10"
    stop: "20.0.0.30"
  - start: "30.1.2.3"
```

The `cidrs` list has been renamed to blocks since we are now dealing
with blocks of IPs not just CIDRs anymore. The old `cidrs` field is
still supported and takes the same elements, so just an alias for
backwards compatibility.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

danehans commented on Documentation/network/lb-ipam.rst:
```diff
@@ -128,8 +154,8 @@ CIDR, we will see the following:
-    blue-pool   false      False         254             25m
-    red-pool    false      True          254             11s
+    blue-pool   false      False         256             25m
+    red-pool    false      True          256             11s
```
This should still be `254` in v1.15 since `.spec.allowFirstLastIPs` defaults to `no`, correct? The same comment for L207.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,21 @@ type CiliumLoadBalancerIPPoolSpec struct {
 	ServiceSelector *slimv1.LabelSelector `json:"serviceSelector"`
-	// CiliumLoadBalancerIPPoolCIDRBlock is a list of CIDRs comprising this IP Pool
+	// AllowFirstLastIPs, if set to `yes`` means that the first and last IPs of each CIDR will be allocatable.
+	// If `no` or undefined, these IPs will be reserved.
```
nit: According to `lb-ipam.rst`, LB-IPAM will ignore this field when specifying a /31-/32 for IPv4 and /131-/132 for IPv6. Not a bad idea to state this in the godocs here.

danehans commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -1070,13 +1081,32 @@ func (ipam *LBIPAM) handleNewPool(ctx context.Context, pool *cilium_api_v2alpha1
+
+		// If AllowFirstLastIPs is no or unspecified, mark the first and last IP as allocated upon range creation.
+		if fromCidr && pool.Spec.AllowFirstLastIPs != cilium_api_v2alpha1.AllowFirstLastIPYes {
+			// TODO: in 1.16 switch from default no to default yes.
```
nit: Not a bad idea to create and link an issue here so this TODO is resolved for v1.16.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -82,11 +92,22 @@ type CiliumLoadBalancerIPPoolSpec struct {
+
+const (
+	AllowFirstLastIPNo  AllowFirstLastIPType = "no"
+	AllowFirstLastIPYes AllowFirstLastIPType = "yes"
```
nit: from my take on the k8s API conventions, alias string values should be capitalized.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -82,11 +92,22 @@ type CiliumLoadBalancerIPPoolSpec struct {
+	// +kubebuilder:validation:Optional
+	Start string `json:"start"`
+	// +kubebuilder:validation:Optional
+	Stop string `json:"stop"`
```
If start/stop represents IP addresses, it's not a bad idea to include validation. For example:
```
// +kubebuilder:validation:Pattern=`((^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$))`
```

Optional fields should be pointers and include the `omitempty` JSON tag.

dylandreimerink replied:
Added the `omitempty`. I am holding off on the validation pattern. While I do understand, the reasoning, such regexes are near impossible to review/verify. They are also not very user friendly since kubectl will then spit back out: `{field} does not match {regex}`, which is huge and equally unreadable there. We have validation in place on the consuming side and report back bad IPs via the status (also not ideal). I would really like to see an improved version of this with https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook at some point.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -69,11 +69,21 @@ type CiliumLoadBalancerIPPoolSpec struct {
+	// Blocks is a list of CIDRs comprising this IP Pool
+	//
+	// +kubebuilder:validation:Optional
+	Blocks []CiliumLoadBalancerIPPoolIPBlock `json:"blocks"`
```
Optional fields should be omitted from JSON serialization by using the `omitempty` tag.

danehans commented on operator/pkg/lbipam/service_store.go:
```diff
@@ -82,7 +83,7 @@ func (sv *ServiceView) isSatisfied() bool {
 			// If reqIP doesn't exist in the list of assigned IPs
 			if slices.IndexFunc(sv.Status.LoadBalancer.Ingress, func(in slim_core_v1.LoadBalancerIngress) bool {
-				return net.ParseIP(in.IP).Equal(reqIP)
+				return netip.MustParseAddr(in.IP).Compare(reqIP) == 0
```
nit: according to the [docs](https://pkg.go.dev/net/netip#MustParseAddr) `MustParseAddr` should only be used in tests.

danehans commented on pkg/bgpv1/manager/route_policy_reconciler_test.go:
```diff
@@ -31,7 +31,7 @@ var (
 		},
 		Spec: v2alpha1api.CiliumLoadBalancerIPPoolSpec{
-			Cidrs: []v2alpha1api.CiliumLoadBalancerIPPoolCIDRBlock{
+			Cidrs: []v2alpha1api.CiliumLoadBalancerIPPoolIPBlock{
```
Should the tests be updated to cover IP ranges too?

dylandreimerink replied:
This test is part of the BGP package, I am honestly not sure why tests from that package are using LB-IPAM in these tests instead of just manually setting `.status.LoadBalancer.ingress[]`. I don't think its necessary to add extensive coverage of LB-IPAM features here.

danehans reviewed (approved):
@dylandreimerink this is great work by you. Since my review comments are mostly nits, I `/approve`.

tommyp1ckles commented on pkg/ipalloc/ipalloc.go:
```diff
@@ -0,0 +1,636 @@
+)
+
+// Allocator allocates IP addresses, `T` is the type of value associated with the IP.
+// The associated value can be used to store metadata about the IP, such as owner, purpose, etc.
```
👍

tommyp1ckles commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumloadbalancerippools.yaml:
```diff
@@ -58,20 +58,47 @@ spec:
             description: Spec is a human readable description for a BGP load balancer
               ip pool.
             properties:
+              allowFirstLastIPs:
```
Would there ever be a need to have this be per block?

dylandreimerink replied:
I don't think so. You can always make 2 pools one with and one without this property set and put CIDRs in the corresponding pools
